crates/quickjs-wasm-rs/src/js_binding/context.rs

```diff
@@ -169,39 +177,71 @@ impl Context {
             }
         };
 
-        unsafe { self.new_callback(wrapped) }
+        self.new_callback(wrapped)
     }
 
-    /// # Safety
-    ///
-    /// The lifetime of values used in `f` are not respected so you need to
-    /// manually ensure any referenced variables live at least as long as this
-    /// context.
-    ///
-    /// The following example will result in undefined behavior:
+    /// Wrap the specified function in a JS function.
     ///
-    /// ```rs
-    /// let bar = "bar".to_string();
-    /// self.create_callback(|_, _, _, _, _| println!("foo: {}", &bar));
-    /// ```
-    pub unsafe fn new_callback<F>(&self, f: F) -> Result<Value>
+    /// See also [wrap_callback].
+    pub fn new_callback<F>(&self, f: F) -> Result<Value>
     where
-        F: FnMut(*mut JSContext, JSValue, c_int, *mut JSValue, c_int) -> JSValue,
+        F: FnMut(*mut JSContext, JSValue, c_int, *mut JSValue, c_int) -> JSValue + 'static,
     {
         let trampoline = build_trampoline(&f);
-        let data = &f as *const _ as *mut c_void as *mut JSValue;
 
-        let raw = JS_NewCFunctionData(self.inner, trampoline, 0, 1, 1, data);
+        unsafe extern "C" fn finalize<F: 'static>(_runtime: *mut JSRuntime, value: JSValue) {
+            drop(Box::from_raw(JS_GetOpaque(
+                value,
+                *CLASSES.lock().unwrap().get(&TypeId::of::<F>()).unwrap(),
+            ) as *mut F))
+        }
```

Comment:
I'd be fine with making it optional.  Like you, I'm primarily concerned with short-lived processes, but I still like having finalizers for cases like this since QuickJS primarily uses reference counting (plus a cycle collector) and can thus provide predictable, deterministic resource disposal.  That can be especially valuable if you have a closure that does more than just free memory when dropped (e.g. closes file descriptors, etc.).

Reply:
I just realized that the `finalize` function will never be called anyway since there's no way to free a `JSValue` that's wrapped in a `Value`.  So once you create a callback using `new_callback`, there's no way to release it.

I left the code in anyway with a comment that it's not (currently) used, just in case it's useful in the future.  Also happy to remove it, if you prefer.